### DISCUSSION TITLE
Using pre-loaded OCSP responses

### DISCUSSION THREAD

**Discussion by tshuli** [MatthiasValvekens/pyHanko#254] [Q&A] [answered]
First of all, thank you for all the amazing work that you've done with this library!

We are trying to use pyHanko for signed pdf documents for the Singapore public sector, which requires the use of LTV. The TSA we are using (Digicert) is on the Adobe Approved Trust List (AATL). 

We are facing an issue where the signatures are not LTV enabled (both for the user as well as the TSA).
<img width="411" alt="Screenshot 2023-04-27 at 4 10 04 PM" src="https://user-images.githubusercontent.com/63710093/234800657-628b3544-1177-4948-a20e-0158c7487456.png">

In our case, we have to retrieve the OCSP responses for the _signing user cert chain_ separately before embedding the validation information, instead of using the inbuilt `allow_fetching` in the `validation_context`, because the OCSP server requires the use of mTLS. For the TSA cert chain, we are not fetching the OCSP responses manually and rely on `allow_fetching`. In the Adobe Acrobat certificate dialogue, it seems that acrobat is falling back to online CRL instead of the using the OCSP responses. 
<img width="405" alt="Screenshot 2023-04-27 at 4 34 37 PM" src="https://user-images.githubusercontent.com/63710093/234806621-a07e4ad3-8e0e-41f8-b967-e31c2fdf96f6.png">

Below is our reference implementation, which uses the [interrupted signing](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#interrupted-signing) flow.

```


validation_context = ValidationContext(
    ocsps=[OCSPResponse.load(base64.b64decode(ocspResponse))
           for ocspResponse in ocsp_responses_b64], #OCSP responses retrieved externally for user cert chain
#certs_asn1 contains the entire chain from user cert to root
    trust_roots=[*certs_asn1,
                 load_cert_from_pemder('src/sign/TSACertificate.cer')], 
    allow_fetching=True
)


def instantiate_external_signer(sig_value: bytes):
    return signers.ExternalSigner(
        signing_cert=user_cert_asn1,
        signature_value=sig_value,
        cert_registry=cert_registry,
        embed_roots=True,
    )

async def createPlaceholder():
    print('creating placeholder')

    with open('/filename', 'rb') as doc:

        w = IncrementalPdfFileWriter(doc)
        fields.append_signature_field(
            w, sig_field_spec=fields.SigFieldSpec(
                'Signature', box=(200, 600, 400, 660)
            )
        )

        ext_signer = instantiate_external_signer(
            sig_value=bytes(256))

        pdf_signer = signers.PdfSigner(
            timestamper=timestamps.HTTPTimeStamper(
                'http://timestamp.digicert.com'),
            signature_meta=signers.PdfSignatureMetadata(
                field_name='Signature',
                md_algorithm='sha256',
                subfilter=SigSeedSubFilter.PADES,
                validation_context=validation_context,
                embed_validation_info=True,
                use_pades_lta=True
            ),
            signer=ext_signer,
            stamp_style=stamp.TextStampStyle(
                stamp_text='This is the signature!'
            )
        )

        prep_digest, tbs_document, output = await pdf_signer.async_digest_doc_for_signing(w)

        signed_attrs = await ext_signer.signed_attrs(
            prep_digest.document_digest, 'sha256', use_pades=True
        )

        psi = tbs_document.post_sign_instructions

        # write prep_digest to stdout
        output = f"Document hash to sign: <{hashlib.sha256(signed_attrs.dump()).hexdigest()}>"

## Serialise and sign externally
## Deserialise once signature is returned and continue

async def finish_signing(prep_digest, psi, output_handle):
    # Here, assume sig_value_bytes is the signed digest of the signed_attrs
    # bytes, obtained from some remote signing service

    ext_signer = instantiate_external_signer(sig_value_bytes)

    sig_cms = await ext_signer.async_sign_prescribed_attributes(
        'sha256', signed_attrs=signed_attrs,
        timestamper=timestamps.HTTPTimeStamper(
            'http://timestamp.digicert.com'),
    )

    await PdfTBSDocument.async_finish_signing(
        output=output_handle, prepared_digest=prep_digest,
        signature_cms=sig_cms,
        post_sign_instr=psi,
        validation_context=validation_context
    )
```

I read in another discussion thread that `trust_roots` should only contain the root CA certificate, and the rest of the certificates should go to `other_certs`. When I adjusted the validation context to:

```
validation_context = ValidationContext(
    ocsps=[OCSPResponse.load(base64.b64decode(ocspResponse))
           for ocspResponse in ocsp_responses_b64],
#certs_asn1 contains the entire chain from user cert to root
    trust_roots=[certs_asn1[-1],
                 load_cert_from_pemder('src/sign/TSACertificate.cer')],  # Must trust the TSA cert. TODO: Shift cert to a more appropriate location
    other_certs=[*certs_asn1[1:-1]],
    allow_fetching=True
)
```

I received this error when calling `PdfTBSDocument.async_finish_signing`

```
[1] [backend] stderr:     await PdfTBSDocument.async_finish_signing(
[1] [backend]   File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/pyhanko/sign/signers/pdf_signer.py", line 2507, in async_finish_signing
[1] [backend] 
[1] [backend] stderr:     await post_sign.post_signature_processing(
[1] [backend]   File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/pyhanko/sign/signers/pdf_signer.py", line 2601, in post_signature_processing
[1] [backend] 
[1] [backend] stderr:     validation.DocumentSecurityStore.supply_dss_in_writer(
[1] [backend]   File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/pyhanko/sign/validation/dss.py", line 464, in supply_dss_in_writer
[1] [backend]     dss.register_vri(
[1] [backend]   File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/pyhanko/sign/validation/dss.py", line 226, in register_vri
[1] [backend] 
[1] [backend] stderr:     crl_refs = set(
[1] [backend]   File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/pyhanko/sign/validation/dss.py", line 142, in _cms_objects_to_streams
[1] [backend]     obj_bytes = obj.dump()
[1] [backend]   File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/asn1crypto/core.py", line 4124, in dump
[1] [backend] 
[1] [backend] stderr:     for index, (field_name, _, params) in enumerate(self._fields):
[1] [backend] 
[1] [backend] stderr: ValueError: not enough values to unpack (expected 3, got 2)
```
Would you have any insight as to what might be the problem?

Thank you so much!

**Comment by MatthiasValvekens** [comment]
Hi @tshuli


Ah, NDI. You're not the first person to come with questions about that... :)

A couple points:

 - The behaviour with the TSA cert in the `trust_roots` is entirely consistent with what I would expect given the rest of your explanation (good job on the investigation, by the way). Since `allow_fetching=True` will still only fetch however much revocation information it needs to validate _a_ path up to a trust root, the result is that effectively nothing will be fetched if the signer's certificate is itself a trust root. So far, nothing out of the ordinary.
 - The value error you're seeing from `asn1crypto` is concerning, though. Could you check which versions of `asn1crypto`, `pyHanko`  and `pyhanko-certvalidator` you're using? This looks to me like something in `asn1crypto`'s global setup logic failed to run. Any chance you're pickling and unpickling objects that contain `asn1crypto` objects? If so: see wbond/asn1crypto#239. The permanent fix hasn't been released yet, but it's been merged and there's a workaround in the issue thread.
 - It's not actually necessary to add the TSA cert itself to `other_certs`, even, since it's a "known quantity". However, it probably doesn't chain up to the same trust root as your NDI cert. There are various ways to deal with that, but if AATL compatibility is your primary concern, the best thing to do would be to put whatever AATL root the TSA cert needs into the `trust_roots` list and any intermediate certs into `other_certs`.
 - Regarding NDI's quirky mTLS-gated OCSP responder: you're right that it's easy enough to manually query it (as you're doing now), but if you really want to, you can pass a custom OCSP fetcher to your `ValidationContext` that is capable of handling NDI certs transparently... Then you can just use `allow_fetching`... :)  It's a bit more involved, though, for obvious reasons.

Hope this helps.

**Reply by tshuli** [reply]
@MatthiasValvekens, thank you for the very detailed response, and for responding so quickly!

To close the loop on the various issues raised:
- Thanks for explaining in detail what `allow_fetching` does
- You are right that the value error was due to serialisation (of `prep_digest, psi, cert_registry`), and it was fixed with the [workaround](https://github.com/wbond/asn1crypto/issues/239) you suggested. These are the library versions I was using:
  - asn1crypto 1.5.1
  - pyHanko 0.17.2
  - pyhanko-certvalidator 0.20.1
- Putting the TSA root cert in `trust_roots` worked for me.
- I'll probably give it a shot at implementing a custom OCSP fetcher for ValidationContext, to ride on the in-built `allow_fetching` functionality.

Once again, thank you for the amazing work on this library and for being so willing to help.

**Comment by tshuli** [comment]
Hi @MatthiasValvekens, thanks again for all your great work and support with this.

A quick follow up to my earlier question - I modified the OCSP fetcher to get past the mTLS gating, and am able to retrieve the OCSP responses for the signing cert chain using `allow_fetching` (this is also verified by logging the ocsp responses). In addition, when I inspect the raw pdf data, I see that the OCSP responses are embedded in my pdf document. However, the CRL responses for the signing cert chain are also being fetched and embedded, and Adobe uses the CRL responses for revocation information for the signing cert chain instead of OCSP. 
<img width="590" alt="Screenshot 2023-05-10 at 1 21 50 PM" src="https://github.com/MatthiasValvekens/pyHanko/assets/63710093/d3ddd15f-62d7-448e-a5d3-d543c55d2637">


For the timestamp cert chain, the ocsp responses are being used as revocation information.

Ideally, I'd like to fetch and use only the OCSP responses as revocation information for the signing cert chain, for a smaller file size. I had a few questions please:

- Does pyhanko still fetch the CRL responses if it is able to fetch the complete set of OCSP responses for the signing cert chain? Or is the fact that pyhanko is fetching the CRL responses, an indication that the OCSP responses are invalid/incomplete?
- If pyhanko fetches both OCSP and CRL responses by default, how do I enforce that only OCSP responses are to be used? 

I also wonder if this might be because I am not configuring the cert registry correctly. This is what I have right now.

- For the signer, I have
```
cert_registry = SimpleCertificateStore()

# Load all the certs from user cert to root cert
for cert_asn in certs_asn1:
    cert_registry.register(
        cert_asn
    )

def instantiate_external_signer(sig_value: bytes):
    return signers.ExternalSigner(
        signing_cert=user_cert_asn1,
        signature_value=sig_value,
        cert_registry=cert_registry,
        embed_roots=True,
    )
``` 
- For Validation Context, I have
```
validation_context = ValidationContext(
    # Both the signing root cert and timestamp root cert are in trust_roots
    trust_roots=[certs_asn1[-1], timestamp_root_cert_asn1],
    allow_fetching=True,
    # Custom fetchers to get past mTLS gating
    fetchers=CustomFetchers
)
```

Thank you once again!

**Reply by MatthiasValvekens** [reply]
Commenting on mobile, so I'll be brief.

CRL fetching should indeed be skipped if there's enough info in the OCSP responses and they can be validated. This is evaluated separately for each step in the chain. Are you on the latest version of pyhanko-certvalidator? There's a recent fix that might impact this.

Also, can you post the debug log output? Maybe there's a hint in there.

**Reply by tshuli** [reply]
@MatthiasValvekens, thank you for responding!

- Yes, I am on `pyhanko-certvalidator` 0.22.0, the latest version.
- I believe the freshness check might be the reason (see logs below). The `pyhanko-certvalidator` [docs](https://pyhanko.readthedocs.io/en/latest/api-docs/pyhanko_certvalidator.html#pyhanko_certvalidator.policy_decl.CertRevTrustPolicy) define 
`Freshness interval. If not specified, this defaults to the distance between thisUpdate and nextUpdate for the given piece of revocation information.`

On further investigation, the OCSP responses I received do have `thisUpdate`, but no `nextUpdate` field. This might be why the freshness check failed. But [RFC 6960](https://www.rfc-editor.org/rfc/rfc6960#section-4.2.2.1) states that ```If nextUpdate is not set, the responder is indicating that newer revocation information is available all the time.```, so this is still a valid OCSP response.

In the SDK, when using the interrupted signing flow, is there any way I can override the freshness check configurations (e.g. when instantiating the signer, or validation context)?



```
File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/pyhanko/sign/signers/pdf_signer.py", line 1753, in _perform_presign_signer_validation
     signer_cert_validation_path = await validator.async_validate_usage(
   File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/pyhanko_certvalidator/__init__.py", line 269, in async_validate_usage
     validated_path = await self.async_validate_path()
   File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/pyhanko_certvalidator/__init__.py", line 136, in async_validate_path
     self._path = candidate_path = await find_valid_path(
   File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/pyhanko_certvalidator/__init__.py", line 53, in find_valid_path
     raise exceptions[0]
   File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/pyhanko_certvalidator/__init__.py", line 36, in find_valid_path
     await async_validate_path(
   File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/pyhanko_certvalidator/validate.py", line 145, in async_validate_path
     return await intl_validate_path(
   File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/pyhanko_certvalidator/validate.py", line 1111, in intl_validate_path
     await _check_revocation(
   File "/Users/<...>/Library/Python/3.9/lib/python/site-packages/pyhanko_certvalidator/validate.py", line 1386, in _check_revocation
     raise InsufficientRevinfoError.from_state(
 pyhanko_certvalidator.errors.InsufficientRevinfoError: The path could not be validated because the intermediate certificate 1 revocation checks failed: OCSP response freshness could not be established
 ```

**Reply by MatthiasValvekens** [reply]
Seems plausible indeed; I'll investigate. In the meantime: does setting a very high time_tolerance help?

**Reply by tshuli** [reply]
1) Setting a very high `time_tolerance` (e.g. 30 days) does not work, I got the same error as above
2) However, setting the revocation policy to `freshness_req_type=FreshnessReqType.MAX_DIFF_REVOCATION_VALIDATION,` with a high `freshness` of 30 days worked
3) Setting the revocation policy to `freshness_req_type=FreshnessReqType.MAX_DIFF_REVOCATION_VALIDATION`, with a low `freshness` of 15 minutes did not work (even though the validation time was within `thisUpdate` + 15 minutes). In this case I got a slightly different error: ```pyhanko_certvalidator.errors.InsufficientRevinfoError: The path could not be validated because the mandatory OCSP check(s) for intermediate certificate 1 failed: OCSP response is not recent enough```

For 3), it seems like `MAX_DIFF_REVOCATION_VALIDATION` should not need to depend on `nextUpdate` based on the documentation:
```
Freshness policy requiring that the validation time, if later than the
    issuance date of the revocation info, be sufficiently close to that
    issuance date.
```

Once again, thank you!

**Reply by MatthiasValvekens** [reply]
Thanks for checking! Yeah, the freshness thing was reworked in a recent-ish release to implement AdES, but I may have overlooked some compat issues with the default/legacy policy.
For (3), are you sure that the constraint is met for the intermediate CA's OCSP response as well, and not just the one for the signer cert? That check should only look at `thisUpdate`.

**Reply by MatthiasValvekens** [reply]
Small update: the case where `nextUpdate` is not set is indeed always rejected by the `DEFAULT` freshness policy. FWIW, I agree that this is overly strict and therefore a bug, but it's less obvious what the "correct" behaviour is in this case in the absence of a properly parameterised freshness policy. I'm not 100% sure what I'll do yet, but I'll probably set it to something like 30 minutes by default.

**Reply by tshuli** [reply]
Thank you @MatthiasValvekens ! A 30 minute default sounds like it might work. In the meantime I'll use case (2) as a workaround.

For (3), I can confirm that the intermediate CA's OCSP response:
- Also only has a `thisUpdate` and no `nextUpdate` field
- Was validated within `thisUpdate` + 15 minutes.

When I checked recovation only for the signer cert:
```
RevocationCheckingPolicy(
            RevocationCheckingRule.OCSP_REQUIRED,
            RevocationCheckingRule.NO_CHECK
        ),
 ```
 I got the same error (this time for the end-entity certificate)
 ```
InsufficientRevinfoError('The path could not be validated because the mandatory OCSP check(s) for the end-entity certificate failed: OCSP response is not recent enough'))
```

For clarity, for cases (2) and (3), I did not set a `time_tolerance`.

**Reply by MatthiasValvekens** [reply, marked as answer]
Hi @tshuli,

Thanks! Can you have another go with `pyhanko-certvalidator==0.23.0`? I introduced the fallback we discussed.

**Reply by tshuli** [reply]
Thank you @MatthiasValvekens ! The fallback works now in `pyhanko-certvalidator==0.23.0` with the default `freshness_req_type`. I really appreciate your willingness to help.

**Reply by tshuli** [reply]
🤔 Unfortunately, I now intermittently get this error
```
pyhanko_certvalidator.errors.InsufficientRevinfoError: The path could not be validated because the mandatory OCSP check(s) for the end-entity certificate failed: OCSP response is too recent
```

I'll be taking a look at `pyhanko-certvalidator` to try and identify the cause of it!

**Reply by MatthiasValvekens** [reply]
Hmm, that one should not appear if `time_tolerance` is set to something reasonable. Usually this is the result of pretty minor clock drift issues.

(As to why this check is even performed: this mirrors the behaviour from the `certvalidator` lib I forked and I kept it around with the goal of supporting things like certificate holds, but actually the logic behind checking those with OCSP is not entirely sound anyway for historical validation, so I might get rid of it entirely...)
